File: src/control/RouteController.java
package control;

import database.DBRoute;
import model.DefaultRoute;
import model.Route;
import model.DeliveryStop;
import model.TransportUnit;

import java.sql.SQLException;
import java.util.*;

/**
 * RouteController
 * Handles all functionality related to routes.
 *
 * @author Dani Sander
 * @version 1.0
 * @since 20-05-15
 */

public class RouteController {

    private DeliveryStopController deliveryStopController;
    private DefaultRouteController defaultRouteController;
    private DefaultDeliveryStopController defaultDeliveryStopController;
    private DBRoute dbRoute;
    private static RouteController instance;
    private ArrayList<Route> routes;
    private double load;

    /**
     * Private constructor for singleton.
     */
    private RouteController() {
        deliveryStopController = DeliveryStopController.getInstance();
        defaultRouteController = DefaultRouteController.getInstance();
        defaultDeliveryStopController = DefaultDeliveryStopController.getInstance();
        try {
            dbRoute = DBRoute.getInstance();
        } catch (ClassNotFoundException | SQLException e) {
            // TODO Auto-generated catch block
            e.printStackTrace();
        }
    }

    /**
     * Singleton method for class.
     *
     * @return instance of class.
     */
    public static RouteController getInstance() {
        if (instance == null) {
            instance = new RouteController();
        }

        return instance;
    }

    /**
     * Imports all routes from database.
     */
    public void importRoutes(Date date) {

        // Sync List
        List<DefaultRoute> listDefaultRoutes = Collections.synchronizedList(
                //Gets a list of all defaultRoutes.
                defaultRouteController.getDefaultRoutes());

        //create a route for each defaultRoute
        listDefaultRoutes.parallelStream().forEach((defaultRoute -> {
            //Creates new routes for each defaultRoute.
            Route route = new Route(defaultRoute, date);

            //Sync then adding
            synchronized (listDefaultRoutes) {
                //Creates Delivery Stops
                deliveryStopController.addDeliveryStops(
                        route,
                        //Gets a list of all stops for the defaultRoute.
                        defaultDeliveryStopController.getDefaultDeliveryStops(defaultRoute)
                );
            }
        }));
    }

    /**
     * Exports all data to database.
     */
    public void exportData() {
        routes.parallelStream().forEach(route -> {
            if (route.getDefaultRoute().isExtraRoute()) {
                defaultRouteController.storeDefaultRoute(route.getDefaultRoute());
            }
            dbRoute.storeRoute(route);
            deliveryStopController.storeDeliveryStops(route);
        });
    }

    
	/**
	 * Finds and returns all overloaded routes.
	 * @return ArrayList containing all overloaded routes.
	 */
	public ArrayList<Route> findOverloadedRoutes() {
		
		//Creates an ArrayList for each overloaded route.
		ArrayList<Route> overloadedRoutes = new ArrayList<>();
		
		//Enters a loop for each route.
        routes.stream().forEach((route) -> {
        	
        	//Variable to increment for each load check.
        	load = 0;
        	
        	//Finds maximum load.
        	double capacity = route.getDefaultRoute().getTrailerType().getCapacity();       	
        	
        	//Enters a loop for each delivery stop.
        	ArrayList<DeliveryStop> stops = route.getStops();       	
        	stops.stream().forEach((stop) -> {
        		
        		//Enters a loop for each transportUnit
        		ArrayList<TransportUnit> transportUnits = stop.getTransportUnits();
        		for(TransportUnit transportUnit:transportUnits) {
        			
        			//Increments load with the transportUnits size.
        			load += transportUnit.getType().getSize();
        		}
        	});
        	
        	//Checks to see if route is overloaded.
        	if (load > capacity) {
        		//Adds overloaded route to ArrayList.
        		overloadedRoutes.add(route);
        	}
        });
		
		//Return list with all overloaded routes.
		return overloadedRoutes;	
	}
	
	/**
	 * Finds and returns all under loaded routes.
	 * @return ArrayList containing all under loaded routes.
	 */
	public ArrayList<Route> findUnderloadedRoutes() {
		
		//Creates an ArrayList for each overloaded route.
		ArrayList<Route> underloadedRoutes = new ArrayList<>();
		
		//Enters a loop for each route.
        routes.stream().forEach((route) -> {
        	
        	//Variable to increment for each load check.
        	load = 0;
        	
        	//Finds maximum load.
        	double capacity = route.getDefaultRoute().getTrailerType().getCapacity();       	
        	
        	//Enters a loop for each delivery stop.
        	ArrayList<DeliveryStop> stops = route.getStops();       	
        	stops.stream().forEach((stop) -> {
        		
        		//Enters a loop for each transportUnit
        		ArrayList<TransportUnit> transportUnits = stop.getTransportUnits();
        		for(TransportUnit transportUnit:transportUnits) {
        			
        			//Increments load with the transportUnits size.
        			load += transportUnit.getType().getSize();
        		}
        	});
        	
        	//Checks to see if route is under loaded.
        	if (load < capacity * 0.8) {
        		//Adds under loaded route to ArrayList.
        		underloadedRoutes.add(route);
        	}
        });
		
		//Return list with all overloaded routes.
		return underloadedRoutes;	
	}

	/**
	 * @return the routes
	 */
	public ArrayList<Route> getRoutes() {
		return routes;
	}
}

File: Control/RouteController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Server.Database;
using System.Collections.Concurrent;
using Model;

namespace Control
{
    public class RouteController
    {
        public DeliveryStopController DeliveryStopCtr { get; private set; }
        public DefaultRouteController DefaultRouteCtr { get; private set; }
        public DefaultDeliveryStopController DefaultDeliveryStopCtr { get; private set; }
        // public LogController LogCtr { get; private set; }
        public DBRoute DbRoute { get; private set; }
        public ConcurrentBag<Route> Routes { get; private set; }
        private static RouteController instance;

        /// <summary>
        /// Private singleton constructor.
        /// </summary>
        private RouteController() {
            DeliveryStopCtr = DeliveryStopController.Instance;
            DefaultRouteCtr = DefaultRouteController.Instance;
            DefaultDeliveryStopCtr = DefaultDeliveryStopController.Instance;
            // LogCtr = LogController.Instance;
            DbRoute = DBRoute.Instance;
            Routes = new ConcurrentBag<Route>();
        }

        /// <summary>
        /// Singleton method. Returns the instance of the class.
        /// </summary>
        /// <returns>Instance of class.</returns>
        public static RouteController Instance {
            get {
                    if (instance == null)
                    instance = new RouteController();
                return instance;
            }
        }

        /// <summary>
        /// Imports all routes from database.
        /// </summary>
        /// <param name="date">Time used in creation of routes.</param>
        public void ImportRoutes(DateTime date) {

            //Loads default routes.
            Routes = new ConcurrentBag<Route>();
            List<DefaultRoute> listDefaultRoutes = DefaultRouteCtr.GetDefaultRoutes();
            ConcurrentBag<DefaultRoute> bagDefaultRoutes = new ConcurrentBag<DefaultRoute>(listDefaultRoutes);
            // LogCtr.StatusLog("Loaded Default Routes");

            //Creates a route for each default route.
            Parallel.ForEach(bagDefaultRoutes, defaultRoute =>
            {
                //Creates the route.
                Route route = new Route(defaultRoute, date, date);
                // LogCtr.StatusLog("Creating new route, based on default route " + defaultRoute.ID);
            
                //Syncronize then add stops.
                lock (bagDefaultRoutes)
                {
                    DeliveryStopCtr.AddDeliveryStops(route, DefaultDeliveryStopCtr.GetDefaultDeliveryStops(defaultRoute));
                }
           
                //Updates log and adds route.
                // LogCtr.StatusLog("Created new route from default route " + defaultRoute.ID);
                Routes.Add(route);
            });
        }

        /// <summary>
        /// Imports all routes from the given .csv files.
        /// </summary>
        /// <param name="date">Time used in creation of routes.</param>
        /// <param name="pathRoutes">Path of .csv file to be imported.</param>
        /// <param name="pathStops">Path of .csv file to be imported.</param>
        /// <param name="pathCustomers">Path of .csv file to be imported.</param>
        public void ImportRoutesFromFile(DateTime date, string pathRoutes, string pathStops, string pathCustomers)
        {
            //Loads default routes.
            Routes = new ConcurrentBag<Route>();
            List<DefaultRoute> listDefaultRoutes = DefaultRouteCtr.GetDefaultRoutes(pathRoutes);
            ConcurrentBag<DefaultRoute> bagDefaultRoutes = new ConcurrentBag<DefaultRoute>(listDefaultRoutes);
            // LogCtr.StatusLog("Loaded Default Routes");

            //Creates a route for each default route.
            Parallel.ForEach(bagDefaultRoutes, defaultRoute =>
            {
                //Creates the route.
                Route route = new Route(defaultRoute, date, date);
                // LogCtr.StatusLog("Creating new route, based on default route " + defaultRoute.ID);

                //Syncronize then add stops.
                lock (bagDefaultRoutes)
                {
                    DeliveryStopCtr.AddDeliveryStops(route, DefaultDeliveryStopCtr.ImportDefaultDeliveryStopsFromFile(defaultRoute, pathStops, pathCustomers));
                }

                //Updates log and adds route.
                // LogCtr.StatusLog("Created new route from default route " + defaultRoute.ID);
                Routes.Add(route);
            });
        }

        /// <summary>
        /// Exports all routes to database. If route contains extra default route,
        /// default route is exported as well.
        /// </summary>
        public void ExportData() {

            //Enters a loop for each route.
            foreach (Route route in Routes)
            {
                //Checks if extra route.
                if (route.DefaultRoute.ExtraRoute)
                {
                    //Saves extra route.
                    DefaultRouteCtr.store(route.DefaultRoute);
                }

                //Stores routes and stops to database.
                DbRoute.storeRoute(route);
                DeliveryStopCtr.StoreDeliveryStops(route);
            
                //Updates log.
                //LogCtr.StatusLog(string.Format("Exported {0} route {1} to database",
                //    route.DefaultRoute.ExtraRoute ? "Extra " : "", 
                //    route.ID
                //    ));
            }
        }
    
        /// <summary>
        /// Finds and returns all overloaded routes.
        /// </summary>
        /// <returns>List of all overloaded Routes.</returns>
        public List<Route> FindOverloadedRoutes() {
        
            //Creates a list of routes.
            List<Route> overloadedRoutes = new List<Route>();

            //Checks if each route is overloaded.
            Parallel.ForEach(overloadedRoutes, route => 
            {
                //Gets variables.
                double load = route.GetLoadForTrailer();
                double capacity = route.GetCapacity();

                //Checks if overloaded.
                if (load > capacity)
                {
                    //Adds the route to list.
                    overloadedRoutes.Add(route);
                }
            });

            //Return list with routes.
            return overloadedRoutes;
        }
    
        /// <summary>
        /// Finds and returns all underloaded routes.
        /// </summary>
        /// <returns>List of all underloaded routes.</returns>
        public List<Route> FindUnderloadedRoutes() {

            //Creates a list of routes.
            List<Route> underloadedRoutes = new List<Route>();

            //Checks if each route is underloaded.
            Parallel.ForEach(underloadedRoutes, route => 
            {
                //Checks if underlaoded.
                if (route.IsUnderloaded())
                {
                    //Adds to list.
                    underloadedRoutes.Add(route);
                }
            });

            //Return list with routes.
            return underloadedRoutes;   
        }


        /// <summary>
        /// Calculates the time for departure.
        /// </summary>
        public void CalcTimeForDeparture() {

            DateTime now = DateTime.Now;
            Parallel.ForEach(Routes, route => 
            {
                route.TimeForDeparture = new DateTime(now.Year, now.Month, now.Day, 0, 0, 0);
            });
        }
    }
}
